File: Python/Memento/Memento.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
'''
 Memento Pattern
 Author: Kei Nakata
 Data: Oct.10.2014
'''
import copy
class Memento(object):
    def __init__(self, message):
        self.message = message

    def getState(self):
        return self.message

    def setState(self, message):
        self.message = message

class Originator(object):
    def __init__(self):
        self.message = "Hello !!"
        self.memento = Memento(self.message)

    def setMessage(self, message):
        self.message = message

    def showMessage(self):
        print self.message

    def createMemento(self):
        self.memento.setState(self.message)
        return copy.deepcopy(self.memento)

    def setMemento(self, memento):
        self.memento = memento
        self.message = self.memento.getState()

class Caretaker(object):
    def __init__(self, originator):
        self.savedata = []
        self.originator = originator

    def rollback(self):
        self.originator.setMemento(self.savedata[-1])
        if len(self.savedata) > 1:
            self.savedata.pop()

    def commit(self):
        buf = self.originator.createMemento()
        self.savedata.append(buf)

if __name__ == '__main__':
    a = Originator()
    b = Caretaker(a)
    a.setMessage("apple")
    a.showMessage()
    b.commit()
    a.setMessage("orange")
    a.showMessage()
    b.commit()
    a.setMessage("banana")
    a.showMessage()
    b.commit()
    b.rollback()
    a.showMessage()
    b.rollback()
    a.showMessage()
    b.rollback()
    a.showMessage()

File: C++/Memento/Memento.cpp
/*
 * Memento pattern
 * Author: reimen
 * Data: Oct.6.2014
 * Without violating encapsulation, capture and externalize an object's internal
 * state so that the object can be restored to this state later
 */
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <stack>
using std::cout;
using std::endl;
using std::string;
using std::stack;

class Memento
{
private:
  string _message = "Hello";
private:
  friend class Originator;
  Memento() = default;
  void setState(const string& message)
  {
    _message = message;
  }
  string getState() const
  {
    return _message;
  }
public:
  virtual ~Memento() = default;
};

class Originator
{
private:
  Memento _memento;
  string _message;
private:
  friend class Caretaker;
  Memento createMemento()
  {
    _memento.setState(_message);
    return _memento;
  }

  void setMemento(const Memento& memento)
  {
    _memento = memento;
    _message = _memento.getState();
  }
public:
  Originator() = default;
  virtual ~Originator() = default;
  void setMessage(const string& message)
  {
    _message = message;
  }

  void printMessage()
  {
    cout << _message << endl;
  }
};

class Caretaker
{
  stack<Memento> _savedata;
  Originator* _originator;
public:
  Caretaker(Originator* originator): _originator(originator) {}
  virtual ~Caretaker() = default;
  void commit()
  {
    Memento buff = (_originator -> createMemento());
    _savedata.push(buff);
  }
  void rollback()
  {
    Memento buff = _savedata.top();
    if(_savedata.size() > 1) _savedata.pop();
    _originator -> setMemento(buff);
  }
};

int main()
{
  Originator a;
  Caretaker b(&a);
  a.setMessage("apple");
  a.printMessage();
  b.commit();
  a.setMessage("banana");
  a.printMessage();
  b.commit();
  a.setMessage("orange");
  a.printMessage();
  b.commit();
  cout << "\n";
  b.rollback();
  a.printMessage();
  b.rollback();
  a.printMessage();
  b.rollback();
  a.printMessage();
  return EXIT_SUCCESS;
}
